pkg/api/norman/customization/aks/handler.go

```diff
@@ -255,28 +257,39 @@ func (h *handler) getCloudCredential(req *http.Request, cap *Capabilities, credI
 }
 
 func (h *handler) clusterCheck(apiContext *types.APIContext, clusterID, cloudCredentialID string) (int, error) {
-	clusterInfo := map[string]interface{}{
-		"id": clusterID,
-	}
-
-	clusterSchema := h.schemas.Schema(&mgmtSchema.Version, client.ClusterType)
-	if err := h.ac.CanDo(v3.ClusterGroupVersionKind.Group, v3.ClusterResource.Name, "update", apiContext, clusterInfo, clusterSchema); err != nil {
-		return httperror.InvalidBodyContent.Status, fmt.Errorf("cluster not found")
+	var (
+		clusters []*v3.Cluster
+		err      error
+	)
+	if clusterID == "" {
+		// If no clusterID is passed, then we check all clusters that the user has access to and are associated to the cloud credential.
+		clusters, err = h.clusterCache.GetByIndex(cluster.ByCloudCredential, cloudCredentialID)
+		if err != nil {
+			return httperror.InvalidBodyContent.Status, err
+		}
+		if len(clusters) == 0 {
+			return httperror.InvalidBodyContent.Status, fmt.Errorf("cloud credential not found")
+		}
+	} else {
+		c, err := h.clusterCache.Get(clusterID)
+		if err != nil {
+			return httperror.ServerError.Status, err
+		}
+		clusters = []*v3.Cluster{c}
 	}
 
-	cluster, err := h.clusterLister.Get("", clusterID)
-	if err != nil {
-		if httperror.IsNotFound(err) {
-			return httperror.InvalidBodyContent.Status, fmt.Errorf("cluster not found")
+	for _, c := range clusters {
+		clusterSchema := h.schemas.Schema(&mgmtSchema.Version, client.ClusterType)
+		if err := h.ac.CanDo(v3.ClusterGroupVersionKind.Group, v3.ClusterResource.Name, "update", apiContext, map[string]interface{}{"id": c.Name}, clusterSchema); err != nil {
+			continue
 		}
-		return httperror.ServerError.Status, err
-	}
 
-	if cluster.Spec.AKSConfig.AzureCredentialSecret != cloudCredentialID {
-		return httperror.InvalidBodyContent.Status, fmt.Errorf("cloud credential not found")
+		if c.Spec.AKSConfig != nil && c.Spec.AKSConfig.AzureCredentialSecret == cloudCredentialID {
```

Comment:
I think these checks, for all the KEv2 providers, can happen prior to the CanDo checks. That way we can skip CanDo if this is the case.

Reply:
I agree that some of this logic can be consolidated.

For now, I reordered the checks as you suggested.